### PR TITLE
Fixes for PLT ABI dynamic linking

### PR DESCRIPTION
**Pull request by arichardson**
This change removes lots of the libc_pic dependencies from rtld and allows rtld to be built for the PLT ABI.
This is done so that no more C-language function pointers are created inside RTLD (an assembler macro is used instead) to avoid adding support for R_MIPS_CHERI_CAPABILITY handling in self-relocation.

I'm not sure the `#ifdef IN_RTLD` changes inside libc are okay but I can't think of a better solution.
A side-effect of this change is that rtld is now noticeably smaller since it no longer pulls in printf() and all the locales stuff + pthread interposing tables.

Finally, I also changed RTLD to be built without debug support and now install /libexec/ld-cheri-elf-debug.so.1 and /libexec/ld-elf-debug.so.1 which has debug support enabled.
Since RTLD direct exec works just fine now, I think this should help our benchmarks to be more similar between static and dynamic linking.

### CODE REVIEW

**Comment by arichardson** [comment]
Size changes for RTLD:

Even if you ignore the DEBUG->nodebug text size still changes from 432471 (old, with debug) -> 358483 (new, with debug, and debug_verbose) -> 206601 (new, no debug)

For MIPS it seems even better 334817 -> 199649 -> 133083.

For some reason CHERI RTLD was also pulling in lots of data which is now 90% less (13112 instead of 123688)
```
Size changes:

   text	   data	    bss	    dec	    hex	filename
 432471	 123688	  22752	 578911	  8d55f	../output-stable/rootfs128/libexec/ld-cheri-elf.so.1
 334817	  19176	  15360	 369353	  5a2c9	../output-stable/rootfs128/libexec/ld-elf.so.1
 358483	  20408	   5472	 384363	  5dd6b	../output/rootfs128/libexec/ld-cheri-elf-debug.so.1
 206601	  13112	   5088	 224801	  36e21	../output/rootfs128/libexec/ld-cheri-elf.so.1
 199649	   3192	   3640	 206481	  32691	../output/rootfs128/libexec/ld-elf-debug.so.1
 133083	   3096	   3656	 139835	  2223b	../output/rootfs128/libexec/ld-elf.so.1
```

Relocation sections only:
```

../output-stable/rootfs128/libexec/ld-cheri-elf.so.1:	file format ELF64-mips

Sections:
Idx Name          Size     VMA          Type
  9 .rel.dyn      0000eb50 00000000000009b0
 16 __cap_relocs  00012610 000000000008bd10 DATA


../output-stable/rootfs128/libexec/ld-elf.so.1:	file format ELF64-mips

Sections:
Idx Name          Size     VMA          Type
  9 .rel.dyn      000020f0 00000000000009d0


../output/rootfs128/libexec/ld-cheri-elf-debug.so.1:	file format ELF64-mips

Sections:
Idx Name          Size     VMA          Type
  9 .rel.dyn      00000010 00000000000009b8
 11 __cap_relocs  0000c580 0000000000005900 DATA


../output/rootfs128/libexec/ld-cheri-elf.so.1:	file format ELF64-mips

Sections:
Idx Name          Size     VMA          Type
  9 .rel.dyn      00000010 00000000000009b0
 11 __cap_relocs  00007e40 0000000000003f70 DATA


../output/rootfs128/libexec/ld-elf-debug.so.1:	file format ELF64-mips

Sections:
Idx Name          Size     VMA          Type
  9 .rel.dyn      00000a50 00000000000009d0


../output/rootfs128/libexec/ld-elf.so.1:	file format ELF64-mips

Sections:
Idx Name          Size     VMA          Type
  9 .rel.dyn      00000a50 00000000000009d0
```

File: libexec/rtld-elf/rtld-libc/Makefile.inc [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+LDFLAGS+=	-Wl,--trace-symbol=_write -Wl,--trace-symbol=write -Wl,--warn-if-file-linked=write.pico
+LDFLAGS+=	-Wl,--trace-symbol=_close  -Wl,--trace-symbol=close -Wl,--warn-if-file-linked=close.pico
+LDFLAGS+=	-Wl,--trace-symbol=_exit  -Wl,--trace-symbol=exit -Wl,--warn-if-file-linked=exit.pico
+LDFLAGS+=	-Wl,--trace-symbol=strerror -Wl,--warn-if-file-linked=strerror.pico
+LDFLAGS+=	-Wl,--trace-symbol=__cleanup -Wl,--trace-symbol=__stderrp
+LDFLAGS+=	-Wl,--trace-symbol=_cleanup -Wl,--warn-if-file-linked=findfp.pico
+.endif
+
+# TODO: avoid this dependency
+LIBADD=		c_pic
```
[the remark this answers is not in the document]
Comment (arichardson) [reply libexec/rtld-elf/rtld-libc/Makefile.inc]: Ah that explains why building rtld for x86 was pulling in lots of stack protector stuff.

**Comment by arichardson** [comment]
I'll try to see if something like this can be upstreamed soon. However, even if not I think it makes sense for us to carry that diff since it speeds up our rtld-cheri-elf quite a lot due to the reduced number of __cap_relocs.

**Comment by arichardson** [comment]
> I spotted a couple minor issue (duplicate line and an over aggressive substitutions of RTLD_ELF_DIR).
> 
> I've got a concern that this isn't a set of changes kib@ will accept upstream and thus we'll end up carrying this diff and the associated conflicts in the rtld Makefiles. It might be worth running the general changes by him before we diverge too much further.

I've rebased this on top of https://reviews.freebsd.org/D20663 (which I'll commit shortly) so there should be no divergence. Is there anything I should fix before commiting?

**Comment by brooksdavis** [comment]
I'm ok with this now.

**Pull request merged**